docs/rfcs/035-safekeeper-dynamic-membership-change.md

```diff
@@ -0,0 +1,491 @@
+updates it in the db if the provided conf generation is higher (the cplane db
+should also store generations for this). Similarly to [`/notify-attach`](https://www.notion.so/neondatabase/Storage-Controller-Control-Plane-interface-6de56dd310a043bfa5c2f5564fa98365), it
+should update db which makes the call successful, and then try to schedule
+`apply_config` if possible, it is ok if not. storage_controller 
+should rate limit calling the endpoint, but likely this won't be needed, as migration
+throughput is limited by `pull_timeline`.
+
+Timeline (branch) creation in cplane should call storage_controller POST
+`tenant/:tenant_id/timeline` like it currently does for sharded tenants.
+Response should be augmented with `safekeeper_conf: Configuration`. The call
+should be retried until succeeds.
+
+Timeline deletion and tenant deletion in cplane should call appropriate
+storage_controller endpoints like it currently does for sharded tenants. The
+calls should be retried until they succeed.
+
+### storage_controller implementation
+
+Current 'load everything on startup and keep in memory' easy design is fine.
+Single timeline shouldn't take more than 100 bytes (it's 16 byte tenant_id, 16
+byte timeline_id, int generation, vec of ~3 safekeeper ids plus some flags), so
+10^6 of timelines shouldn't take more than 100MB.
+
+Similar to pageserver attachment Intents storage_controller would have in-memory
+`MigrationRequest` (or its absense) for each timeline and pool of tasks trying
+to make these request reality; this ensures one instance of storage_controller
+won't do several migrations on the same timeline concurrently. In the first
+version it is simpler to have more manual control and no retries, i.e. migration
+failure removes the request. Later we can build retries and automatic
+scheduling/migration. `MigrationRequest` is
+```
+enum MigrationRequest {
+    To(Vec<NodeId>),
+    FinishPending,
+}
+```
+
+`FinishPending` requests to run the procedure to ensure state is clean: current
+configuration is not joint and majority of safekeepers are aware of it, but do
+not attempt to migrate anywhere. If current configuration fetched on step 1 is
+not joint it jumps to step 7. It should be run at startup for all timelines (but
+similarly, in the first version it is ok to trigger it manually).
+
+#### Schema
+
+`safekeepers` table mirroring current `nodes` should be added, except that for
+`scheduling_policy` field (seems like `status` is a better name for it): it is enough
+to have at least in the beginning only 3 fields: 1) `active` 2) `unavailable` 3)
+`decomissioned`.
+
+`timelines` table:
+```
+table! {
+    // timeline_id is primary key
+    timelines (timeline_id) {
+        timeline_id -> Varchar,
+        tenant_id -> Varchar,
+        generation -> Int4,
+        sk_set -> Jsonb, // list of safekeeper ids
+        new_sk_set -> Nullable<Jsonb>, // list of safekeeper ids, null if not join conf
+        cplane_notified_generation -> Int4,
+    }
```

Comment:
Probably not a matter for this RFC but asking in case there is something I know know about the system/postgres here:

> `timelines (timeline_id) {`

I don't think this is unique enough. does cplane consider `timeline_id` as unique? I would had expected that to be `(tenant_id, timeline_id)`, or perhaps other way around.

Does it matter for queries you had been thinking we'll need?

> `timeline_id -> Varchar,`

using varchar for the ids makes no sense to me, wouldn't the `bytea` be the natural way to store these 16 byte values? or is there an idea that we would be storing something which requires it to be hex string plus something more?

> `sk_set -> JsonB`
> `new_sk_set -> Nullable<JsonB>`

similarly here, is the idea to store more than the vec of identifiers given you've written JsonB instead of something looking like a postgres array?

Reply:
> I don't think this is unique enough. does cplane consider timeline_id as unique?

I believe yes. But including tenant_id in primary key is reasonable, let's do that indeed, changed.

> using varchar for the ids makes no sense to me, wouldn't the bytea be the natural way to store these 16 byte values?

I just copied it from existing tenant_id in `schema.rs`. I think either is ok, we can store directly binary value as bytea or hex string representation, it just must be unique.

> similarly here, is the idea to store more than the vec of identifiers given you've written JsonB instead of something looking like a postgres array?

Right, it is only a vec. Array would work as well, I don't have any preferences here, put jsonb because it is slightly more modern. see e.g.
https://stackoverflow.com/questions/28929834/does-jsonb-make-postgresql-arrays-useless

Reply:
switched to array after chatting